internal/mode/static/handler.go

```diff
@@ -92,3 +118,81 @@ func (h *eventHandlerImpl) updateNginx(ctx context.Context, conf dataplane.Confi
 
 	return nil
 }
+
+// updateControlPlane updates the control plane configuration with the given user spec.
+// If any fields are not set within the user spec, the default configuration values are used.
+func (h *eventHandlerImpl) updateControlPlane(ctx context.Context, cfg *nkgAPI.NginxGateway) {
+	// build up default configuration
+	defaultLogLevel := nkgAPI.ControllerLogLevelInfo
+	controlConfig := nkgAPI.NginxGatewaySpec{
+		Logging: &nkgAPI.Logging{
+			Level: &defaultLogLevel,
+		},
+	}
+
+	updateControlPlane := func() error {
+		// by marshaling the user config and then unmarshaling on top of the default config,
+		// we ensure that any unset user values are set with the default values
+		if cfg != nil {
+			cfgBytes, err := json.Marshal(cfg.Spec)
+			if err != nil {
+				return fmt.Errorf("error marshaling control config: %w", err)
+			}
+
+			if err := json.Unmarshal(cfgBytes, &controlConfig); err != nil {
+				return fmt.Errorf("error unmarshaling control config: %w", err)
+			}
+		} else {
+			msg := "NginxGateway configuration was deleted; using defaults"
+			h.cfg.logger.Error(nil, msg)
+			h.cfg.eventRecorder.Event(
+				&nkgAPI.NginxGateway{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: h.cfg.controlConfigNSName.Namespace,
+						Name:      h.cfg.controlConfigNSName.Name,
+					},
+				},
+				apiv1.EventTypeWarning,
+				"ResourceDeleted",
+				msg,
+			)
+		}
+
+		// set the log level
+		level, err := zapcore.ParseLevel(string(*controlConfig.Logging.Level))
+		if err != nil {
+			return fmt.Errorf("error parsing log level string: %w", err)
+		}
+		h.cfg.atomicLevel.SetLevel(level)
+
+		return nil
+	}
+
+	var cond []conditions.Condition
+	if err := updateControlPlane(); err != nil {
+		msg := "Failed to update control plane configuration"
+		h.cfg.logger.Error(err, msg)
+		h.cfg.eventRecorder.Eventf(
+			cfg,
+			apiv1.EventTypeWarning,
+			"FailedUpdate",
+			"%s; "+msg,
```

Comment:
This was mainly because I didn't want to call `fmt.Sprintf` to include the `msg` in this message, since it gets called again by Eventf to add the error itself. But if that's clearer, I can do that.

Reply:
I see. This is fine, I was just curious since you are using Eventf.